mithril-aggregator/src/artifact_builder/cardano_database.rs

```diff
@@ -0,0 +1,207 @@
+use std::path::{Path, PathBuf};
+
+use anyhow::{anyhow, Context};
+use async_trait::async_trait;
+use semver::Version;
+
+use mithril_common::{
+    entities::{
+        ArtifactsLocations, CardanoDatabase, CardanoDbBeacon, Certificate, CompressionAlgorithm,
+        ProtocolMessagePartKey, SignedEntityType,
+    },
+    StdResult,
+};
+
+use crate::artifact_builder::ArtifactBuilder;
+
+pub struct CardanoDatabaseArtifactBuilder {
+    db_directory: PathBuf, // TODO: temporary, will be accessed through another dependency instead of direct path.
+    cardano_node_version: Version,
+    compression_algorithm: CompressionAlgorithm,
+}
+
+impl CardanoDatabaseArtifactBuilder {
+    pub fn new(
+        db_directory: PathBuf,
+        cardano_node_version: &Version,
+        compression_algorithm: CompressionAlgorithm,
+    ) -> Self {
+        Self {
+            db_directory,
+            cardano_node_version: cardano_node_version.clone(),
+            compression_algorithm,
+        }
+    }
+}
+
+#[async_trait]
+impl ArtifactBuilder<CardanoDbBeacon, CardanoDatabase> for CardanoDatabaseArtifactBuilder {
+    async fn compute_artifact(
+        &self,
+        beacon: CardanoDbBeacon,
+        certificate: &Certificate,
+    ) -> StdResult<CardanoDatabase> {
+        let merkle_root = certificate
+            .protocol_message
+            .get_message_part(&ProtocolMessagePartKey::CardanoDatabaseMerkleRoot)
+            .ok_or(anyhow!(
+                "Can not find CardanoDatabaseMerkleRoot protocol message part in certificate"
+            ))
+            .with_context(|| {
+                format!(
+                    "Can not compute CardanoDatabase artifact for signed_entity: {:?}",
+                    SignedEntityType::CardanoDatabase(beacon.clone())
+                )
+            })?;
+        let total_db_size_uncompressed = compute_uncompressed_database_size(&self.db_directory)?;
+
+        let cardano_database = CardanoDatabase::new(
+            merkle_root.to_string(),
+            beacon,
+            total_db_size_uncompressed,
+            ArtifactsLocations::default(), // TODO: temporary default locations, will be injected in next PR.
+            self.compression_algorithm,
+            &self.cardano_node_version,
+        );
+
+        Ok(cardano_database)
+    }
+}
+
+// Return the sum of the files size contained in the subdirectories 'immutable', 'ledger' and 'volatile'.
+fn compute_uncompressed_database_size(db_directory: &Path) -> StdResult<u64> {
+    let subdirs = ["immutable", "ledger", "volatile"];
+
+    let mut total_db_size_uncompressed = 0;
+    for subdir in subdirs {
+        let dir_path = db_directory.join(subdir);
+
+        total_db_size_uncompressed += get_directory_size(&dir_path)
+            .with_context(|| format!("Failed to read metadata for directory: {:?}", dir_path))?;
+    }
+
+    Ok(total_db_size_uncompressed)
+}
+
+fn get_directory_size(path: &Path) -> StdResult<u64> {
```

Comment:
It could work like this:

```
fn compute_fs_entry_size(path: &Path) -> StdResult<u64> {
    if path.is_file() {
        let metadata = std::fs::metadata(path)
            .with_context(|| format!("Failed to read metadata for file: {:?}", path))?;

        return Ok(metadata.len());
    }

    if path.is_dir() {
        let entries = std::fs::read_dir(path)
            .with_context(|| format!("Failed to read directory: {:?}", path))?;
        let mut directory_size = 0;
        for entry in entries {
            let path = entry
                .with_context(|| format!("Failed to read directory entry in {:?}", path))?
                .path();
            directory_size += compute_fs_entry_size(&path)?;
        }

        return Ok(directory_size);
    }

    Ok(0)
}
```

What do you think?

Reply:
Ok, it's fine